wled00/mqtt.cpp

```diff
@@ -195,4 +270,145 @@ bool initMqtt()
   mqtt->connect();
   return true;
 }
+
+void otaUpdate(String url)
+{
+  DEBUG_PRINT(F("OTA update from URL: "));
+  DEBUG_PRINTLN(url);
+
+  // make client for HTTP request
+  HTTPClient http;
+  http.begin(url);
+  http.setTimeout(3000000); // 5 minute timeout, may change
+
+  // do a get request to get the update binary
+  int httpCode = http.GET();
+  // make sure the request was successful
+  if (httpCode != HTTP_CODE_OK)
+  {
+    DEBUG_PRINT(F("HTTP GET failed, code: "));
+    DEBUG_PRINTLN(httpCode);
+    http.end();
+    return;
+  }
+
+  // disable the watchdog
+  WLED::instance().disableWatchdog();
+  otaInProgress = true; // I've tried both with and without this, neither seems to work
+
+  // get the size of the update
+  int len = http.getSize();
+  DEBUG_PRINT(F("Update size: "));
+  DEBUG_PRINTLN(len);
+
+  // make a buffer for reading
+  WiFiClient *stream = http.getStreamPtr();
+
+  DEBUG_PRINTLN("Got stream");
+
+  // Initialize Update
+  if (!Update.begin(len))
+  {
+    DEBUG_PRINTLN(F("Update.begin failed, most likely not enough space"));
+    http.end();
+    WLED::instance().enableWatchdog();
+    otaInProgress = false;
+    return;
+  }
+
+  DEBUG_PRINTLN("Update.begin succeeded");
+  DEBUG_PRINTLN("Is the stream null?");
+  DEBUG_PRINTLN(stream == nullptr);
+  DEBUG_PRINT(F("Free Heap: "));
+  DEBUG_PRINTLN(ESP.getFreeHeap());
+
+  // write the update to the device
+  size_t written = 0;
+  int bufferSize = 512;
+  uint8_t buffer[bufferSize];
+  // size_t written = Update.writeStream(*stream);
+  while (http.connected() && written < len)
+  {
+    if (stream->available())
+    {
+      int bytesRead = stream->readBytes(buffer, bufferSize);
+      if (bytesRead == 0)
+      {
+        DEBUG_PRINTLN("No bytes read");
+      }
+      written += Update.write(buffer, bytesRead);
+      DEBUG_PRINT("Bytes written: ");
+      DEBUG_PRINTLN(written);
+      if (ESP.getFreeHeap() < 80000)
+      {
+        DEBUG_PRINT(F("Free Heap below 80000: "));
+        DEBUG_PRINTLN(ESP.getFreeHeap());
+      }
+      if (http.connected() != 1)
+      {
+        DEBUG_PRINT("http Connection status: ");
+        DEBUG_PRINTLN(http.connected());
+      }
+      if (WiFi.status() != 3)
+      {
+        DEBUG_PRINT("Wifi status: ");
+        DEBUG_PRINTLN(WiFi.status());
+      }
+    }
+    else
+    {
+      DEBUG_PRINTLN("No bytes available");
+    }
+    delay(10);
+  }
+
+  DEBUG_PRINTLN("Wrote stream");
+
+  // check if the update was successful
+  if (written == len)
+  {
+    DEBUG_PRINTLN(F("Written to flash successfully"));
+  }
+  else
+  {
+    DEBUG_PRINT(F("Update failed, only wrote : "));
```

Comment:
Most users will not be running a debug build, so I think that at least the key messages should also be sent as m via MQTT. 

It looks like you started doing this and then just swapped to debug only

Reply:
That's fair. I wanted to avoid sending mqtt messages at any point in this process until I got it working first just to make sure that isn't causing any issues.